Lib/test/test_annotationlib.py

```diff
@@ -0,0 +1,647 @@
+            set: {a + b},  # single element because order is not guaranteed
+            dict: {a + b: c + d, "key": e + g},
+            list: [a, b, c],
+            tuple: (a, b, c),
+            slice: (a[b:c], a[b:c:d], a[:c], a[b:], a[:], a[::d], a[b::d]),
+            extended_slice: a[:, :, c:d],
+            unpack1: [*a],
+            unpack2: [*a, b, c],
+        ):
+            pass
+
+        anno = annotationlib.get_annotations(f, format=annotationlib.Format.SOURCE)
+        self.assertEqual(
+            anno,
+            {
+                "nested": "list[Annotated[set[int], 'set of ints', 4j]]",
+                "set": "{a + b}",
+                "dict": "{a + b: c + d, 'key': e + g}",
+                "list": "[a, b, c]",
+                "tuple": "(a, b, c)",
+                "slice": "(a[b:c], a[b:c:d], a[:c], a[b:], a[:], a[::d], a[b::d])",
+                "extended_slice": "a[:, :, c:d]",
+                "unpack1": "[*a]",
+                "unpack2": "[*a, b, c]",
+            },
+        )
+
+    def test_unsupported_operations(self):
+        format_msg = "Cannot stringify annotation containing string formatting"
+
+        def f(fstring: f"{a}"):
+            pass
+
+        with self.assertRaisesRegex(TypeError, format_msg):
+            annotationlib.get_annotations(f, format=annotationlib.Format.SOURCE)
+
+        def f(fstring_format: f"{a:02d}"):
+            pass
+
+        with self.assertRaisesRegex(TypeError, format_msg):
+            annotationlib.get_annotations(f, format=annotationlib.Format.SOURCE)
+
+
+class TestForwardRefClass(unittest.TestCase):
+    def test_special_attrs(self):
+        # Forward refs provide a different introspection API. __name__ and
+        # __qualname__ make little sense for forward refs as they can store
+        # complex typing expressions.
+        fr = annotationlib.ForwardRef("set[Any]")
+        self.assertFalse(hasattr(fr, "__name__"))
+        self.assertFalse(hasattr(fr, "__qualname__"))
+        self.assertEqual(fr.__module__, "annotationlib")
+        # Forward refs are currently unpicklable once they contain a code object.
+        fr.__forward_code__  # fill the cache
+        for proto in range(pickle.HIGHEST_PROTOCOL + 1):
+            with self.assertRaises(TypeError):
+                pickle.dumps(fr, proto)
+
+
+class TestGetAnnotations(unittest.TestCase):
+    def test_builtin_type(self):
+        self.assertEqual(annotationlib.get_annotations(int), {})
+        self.assertEqual(annotationlib.get_annotations(object), {})
+
+    def test_custom_metaclass(self):
+        class Meta(type):
+            pass
+
+        class C(metaclass=Meta):
+            x: int
+
+        self.assertEqual(annotationlib.get_annotations(C), {"x": int})
+
+    def test_missing_dunder_dict(self):
+        class NoDict(type):
+            @property
+            def __dict__(cls):
+                raise AttributeError
+
+        class C1(metaclass=NoDict):
+            a: int
```

Comment:
This one still has a `__dict__`, so you should instead use something like:

```python
class NoDict(type):
    def __getattr__(self, item):
        raise AttributeError
    
    def __getattribute__(self, item):
        if item == '__dict__':
            raise AttributeError
        return super().__getattribute__(item)
```

Reply:
`vars(C1)` raises an error here already:

```
>>> class NoDict(type):
...     @property
...     def __dict__(cls):
...         raise AttributeError
... 
... class C1(metaclass=NoDict):
...     a: int
... 
>>> vars(C1)
Traceback (most recent call last):
  File "<python-input-1>", line 1, in <module>
    vars(C1)
    ~~~~^^^^
TypeError: vars() argument must have __dict__ attribute
```

However, the tests I added don't yet cover cases where we call `vars()` on the "owner", I'll work on that later.

Reply:
Oh! this behaviour changed between two versions actually! (I incorrectly checked it with 3.12.2 actually and not 3.14!)

Reply:
> Oh! this behaviour changed between two versions actually! (I incorrectly checked it with 3.12.2 actually and not 3.14!)

Um, that sounds bad. We should try to bisect that and make sure it was deliberate.

Reply:
Oh no false alarm. How come I got no error yesterday but I got one today? maybe it's an even older version. Let me verify again.

Reply:
Ah ! I know, actually I was trying to make the error on the *type* of `B`, namely `vars(type(B))` and not on `B` itself. However, it should be possible to have *owner = NoDict* or is it not possible?

Reply:
This actually made me find an even more basic bug:

```
>>> class X(type):
...     a: int
...     
>>> class Y(metaclass=X):
...     b: str
...     
>>> Y.__annotations__
{'a': <class 'int'>}
```

Reply:
Posted https://discuss.python.org/t/pep-749-implementing-pep-649/54974/28 about this, I think I'll want to deal with it in a separate PR (since the solution likely involves changes to #119361).

Reply:
For this PR I won't handle the metaclass case yet; still thinking about how to handle that.